File: leetcode_python/Math/integer-to-english-words.py
"""

273. Integer to English Words
Hard

Convert a non-negative integer num to its English words representation.

 

Example 1:

Input: num = 123
Output: "One Hundred Twenty Three"
Example 2:

Input: num = 12345
Output: "Twelve Thousand Three Hundred Forty Five"
Example 3:

Input: num = 1234567
Output: "One Million Two Hundred Thirty Four Thousand Five Hundred Sixty Seven"
 

Constraints:

0 <= num <= 231 - 1

"""

# V0

# V1
# https://leetcode.com/problems/integer-to-english-words/discuss/70643/Short-Python
class Solution(object):
    def numberToWords(self, num):
        """
        :type num: int
        :rtype: str
        """
        A = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", \
             "Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", \
             "Sixteen", "Seventeen", "Eighteen", "Nineteen"]
        B = ["Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
        C = ["Billion", "Million", "Thousand"]
        def trans(n):
            return [] if not n else [A[n // 100], "Hundred"] + trans(n%100) if n >= 100 \
                   else [A[n]] if n < 20 else [B[(n//10)-2]] + trans(n%10)
                    
        res, N = [], [10**9, 10**6, 1000, 1]
        for i in range(4):
            digit, num = divmod(num, N[i])
            if digit: res += trans(digit) + C[i:i+1]
        return " ".join(res) or "Zero"

# V1'
class Solution:
    def numberToWords(self, num):
        tens = [
            "",
            "Ten",
            "Twenty",
            "Thirty",
            "Forty",
            "Fifty",
            "Sixty",
            "Seventy",
            "Eighty",
            "Ninety"
        ]
        
        teens = {
            10: 'Ten',
            11: 'Eleven',
            12: 'Twelve',
            13: 'Thirteen',
            14: 'Fourteen',
            15: 'Fifteen',
            16: 'Sixteen',
            17: 'Seventeen',
            18: 'Eighteen',
            19: 'Nineteen',
        }        
        
        digits = [
            "",
            "One",
            "Two",
            "Three",
            "Four",
            "Five",
            "Six",
            "Seven",
            "Eight",
            "Nine"
        ]
        
        def parse(num):
            if num >= 1000000000:
                return parse(num // 1000000000).strip() + " Billion " + parse(num % 1000000000).strip()
            elif num >= 1000000:
                return parse(num // 1000000).strip() + " Million " + parse(num % 1000000).strip()
            elif num >= 1000:
                return parse(num // 1000).strip() + " Thousand " + parse(num % 1000).strip()
            elif num >= 100:
                return digits[num // 100] + " Hundred " + parse(num % 100).strip()
            elif num >= 20:
                return tens[num//10] + " " + parse(num % 10).strip()
            elif num >= 10 and num < 20:
                return teens[num]             
            else:
                return digits[num]
        
        if num == 0:
            return "Zero"
        return parse(num).strip()

# V1''
# (TLE)
# https://leetcode.com/problems/integer-to-english-words/discuss/213494/Python-solution
class Solution:
    def numberToWords(self, num):
        """
        :type num: int
        :rtype: str
        """
        def helper(num):
            if num < 10:
                return dic1[num]
            elif 10 <= num < 20:
                return dic3[num]
            elif num < 100:
                res = []
                q, num = divmod(num, 10)
                res.append(dic2[q*10])
                if num > 0:
                    res.append(" "+dic1[num])
                return "".join(res)
            elif num < 1000:
                q, num = divmod(num, 100)
                if num == 0:
                    return dic1[q]+" "+"Hundred"
                else:
                    return dic1[q]+" "+"Hundred"+" "+helper(num)
            elif num < 1000000:
                q, num = divmod(num, 1000)
                if num == 0:
                    return helper(q)+" "+"Thousand"
                else:
                    return helper(q)+" "+"Thousand"+" "+helper(num)
            elif num < 1000000000:
                q, num = divmod(num, 1000000)
                if num == 0:
                    return helper(q)+" "+"Million"
                else:
                    return helper(q)+" "+"Million"+" "+helper(num)
            else:
                q, num = divmod(num, 1000000000)
                if num == 0:
                    return helper(q)+" "+"Billion"
                else:
                    return helper(q)+" "+"Billion"+" "+helper(num)  
                
        if num == 0:
            return "Zero"
        
        dic1 = {9:"Nine", 8:"Eight", 7:"Seven", 6:"Six", 
                5:"Five", 4:"Four", 3:"Three", 2:"Two", 1:"One"}
        dic2 = {90:"Ninety", 80:"Eighty", 70:"Seventy", 60:"Sixty", 
                50:"Fifty", 40:"Forty", 30:"Thirty", 20:"Twenty"} 
        dic3 = {19:"Nineteen", 18:"Eighteen", 17:"Seventeen", 
                16:"Sixteen", 15:"Fifteen", 14:"Fourteen", 13:"Thirteen", 
                12:"Twelve", 11:"Eleven", 10:"Ten"}
        
        return helper(num)

# V2

File: leetcode_java/src/main/java/LeetCodeJava/Math/IntegerToEnglishWords.java
package LeetCodeJava.Math;

// https://leetcode.com/problems/integer-to-english-words/

import java.util.ArrayList;
import java.util.List;

/**
 *  273. Integer to English Words
 *  Hard
 *
 *  Convert a non-negative integer num to its English words representation.
 *
 *  Example 1:
 *
 *  Input: num = 123
 *  Output: "One Hundred Twenty Three"
 *
 *  Example 2:
 *
 *  Input: num = 12345
 *  Output: "Twelve Thousand Three Hundred Forty Five"
 *
 *  Example 3:
 *
 *  Input: num = 1234567
 *  Output: "One Million Two Hundred Thirty Four Thousand Five Hundred Sixty Seven"
 *
 *  Constraints:
 *
 *  0 <= num <= 2^31 - 1
 */
public class IntegerToEnglishWords {

    private static final String[] BELOW_20 = {
            "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
            "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen",
            "Seventeen", "Eighteen", "Nineteen"
    };

    private static final String[] TENS = {
            "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
    };

    private static final String[] GROUPS = {"", "Thousand", "Million", "Billion"};

    // V0
    // IDEA: split the number into 3-digit groups, spell each group, append its scale word
    /**
     * time = O(1)   (at most 4 groups, each of constant size; O(log n) in digit count)
     * space = O(1)
     */
    public String numberToWords(int num) {
        if (num == 0) {
            return "Zero";
        }
        List<String> parts = new ArrayList<>();
        int groupIdx = 0;
        while (num > 0) {
            int group = num % 1000;
            if (group != 0) {
                List<String> cur = new ArrayList<>();
                helper(group, cur);
                if (groupIdx > 0) {
                    cur.add(GROUPS[groupIdx]);
                }
                // groups are produced from the lowest scale upward -> prepend
                parts.addAll(0, cur);
            }
            num /= 1000;
            groupIdx++;
        }
        StringBuilder sb = new StringBuilder();
        for (String p : parts) {
            if (sb.length() > 0) {
                sb.append(' ');
            }
            sb.append(p);
        }
        return sb.toString();
    }

    // spell a value in [1, 999]
    private void helper(int n, List<String> out) {
        if (n == 0) {
            return;
        }
        if (n >= 100) {
            out.add(BELOW_20[n / 100]);
            out.add("Hundred");
            helper(n % 100, out);
        } else if (n >= 20) {
            out.add(TENS[n / 10]);
            helper(n % 10, out);
        } else {
            out.add(BELOW_20[n]);
        }
    }
}
